src/vcon/MarketGovernance.sol

```diff
@@ -0,0 +1,699 @@
+    /// alternatively, pack venueLastRecordedSharePrice,
+    /// venueLastRecordedProfit and venueVconDeposited into a single slot for gas optimization
+
+    /// @notice last recorded profit index per venue
+    mapping(address => uint128) public venueLastRecordedProfit;
+
+    /// @notice last recorded share price of a venue in VCON
+    /// starts off at 1e18
+    mapping(address => uint128) public venueLastRecordedSharePrice;
+
+    /// @notice total vcon deposited per venue
+    mapping(address => uint256) public venueVconDeposited;
+
+    /// ---------- Per Venue User Profit Tracking ----------
+
+    /// @dev convention for all double nested address mappings is key (venue -> user) -> value
+
+    /// @notice record of VCON index when user joined a given venue
+    mapping(address => mapping(address => uint256))
+        public venueUserStartingVconProfit;
+
+    /// @notice record how much VCON a user deposited in a given venue
+    mapping(address => mapping(address => uint256))
+        public venueUserDepositedVcon;
+
+    /// @param _core reference to core
+    constructor(address _core) CoreRefV2(_core) {}
+
+    /// @notice permissionlessly initialize a venue
+    /// required to be able to utilize a given PCV Deposit in market governance
+    function initializeVenue(address venue) external globalLock(1) {
+        require(pcvOracle().isVenue(venue), "MarketGovernance: invalid venue");
+
+        uint256 startingLastRecordedProfit = venueLastRecordedProfit[venue];
+        require(
+            startingLastRecordedProfit == 0,
+            "MarketGovernance: profit already recorded"
+        );
+        require(
+            venueLastRecordedSharePrice[venue] == 0,
+            "MarketGovernance: venue already has share price"
+        );
+
+        IPCVDepositV2(venue).accrue();
+
+        uint256 endingLastRecordedProfit = IPCVDepositV2(venue)
+            .lastRecordedProfit();
+
+        venueLastRecordedProfit[venue] = endingLastRecordedProfit.toUint128();
+        venueLastRecordedSharePrice[venue] = Constants
+            .ETH_GRANULARITY
+            .toUint128();
+    }
+
+    /// TODO update pcv oracle to cache total pcv so getAllPCV isn't needed to figure
+    /// out if weights are correct
+
+    /// ---------- Permissionless User PCV Allocation Methods ----------
+
+    /// any losses or gains are applied to venueLastRecordedSharePrice
+    ///
+    ///
+    /// user deposits
+
+    /// @notice a user can get slashed up to their full VCON stake for entering
+    /// a venue that takes a loss.
+    /// @param amountVcon to stake on destination
+    /// @param amountPcv to move from source to destination
+    /// @param source address to pull funds from
+    /// @param destination address to accrue rewards to, and send funds to
+    /// @param swapper address that swaps asset types between src and dest if needed
+    function stake(
+        uint256 amountVcon,
```

Comment:
make all remaining params an array

Reply:
I guess if you add multicall that's not needed